fix warning in PyPIConGPU version tests

SimeonEhrig opened the pull request:
- If the script `requirements_txt_modifier.py` should modify a dependency in a requirements.txt, which does not exist, it throws a warning because this is unexpected behavior. Because warnings are hard to detect in the CI, change it to an error.
- Implement a function to ignore specific environment variables which are set to change the version in a requirements.txt.

Example: https://gitlab.com/hzdr/crp/picongpu/-/jobs/7789974598

SimeonEhrig commented:
@BrianMarre please review

BrianMarre reviewed (approved):
Looks good!

psychocoderHPC commented:
@SimeonEhrig Does it mean the test you pointed to will now fail with this PR?

SimeonEhrig commented:
The PR also contains the fix for the current setup. But in future, if the setup will be changed, hopefully yes.

chillenzer commented on share/ci/install/requirements_txt_modifier.py:
Please also adjust the docstring and add the further arguments [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/5d6a9a21cb7c59d4228e79e56c35f5ab144b1ecf/share/ci/install/requirements_txt_modifier.py#L18) (sorry, can't comment so far away from your changes).

SimeonEhrig replied:
This is the help text, which you get, if you run the script without arguments. Normally it is white.

chillenzer replied:
You are correct. But I think you commented on the wrong comment here, right?

chillenzer replied:
I think you never responded to this one. It would be nice to add a line or two after line 19.

chillenzer commented on share/ci/install/pypicongpu.sh:
```diff
@@ -43,8 +43,8 @@ micromamba activate pypicongpu
 MODIFIED_REQUIREMENT_TXT_PYPICONGPU=$CI_PROJECT_DIR/lib/python/picongpu/pypicongpu/modified_requirements.txt
-python3 $CI_PROJECT_DIR/share/ci/install/requirements_txt_modifier.py $CI_PROJECT_DIR/lib/python/picongpu/picmi/requirements.txt $MODIFIED_REQUIREMENT_TXT_PICMI
-python3 $CI_PROJECT_DIR/share/ci/install/requirements_txt_modifier.py $CI_PROJECT_DIR/lib/python/picongpu/pypicongpu/requirements.txt $MODIFIED_REQUIREMENT_TXT_PYPICONGPU
+python3 $CI_PROJECT_DIR/share/ci/install/requirements_txt_modifier.py PYPIC_DEP_VERSION_referencing PYPIC_DEP_VERSION_jsonschema $CI_PROJECT_DIR/lib/python/picongpu/picmi/requirements.txt $MODIFIED_REQUIREMENT_TXT_PICMI
```
Just for my information: Where do these environment variables come from?

SimeonEhrig replied:
The script `pypicongpu_generator.py` generates the jobs dynamically during the CI pipeline lifetime. It set also the environment variables.

chillenzer commented on share/ci/install/requirements_txt_modifier.py:
```diff
@@ -53,21 +53,27 @@ def cs(text: str, color: str):
+              "and the path of the output requirements.txt as the last argument.")
+        print("Set the name of the environment variables to ignore "
+              "before the path of the input requirements.txt")
+        sys.exit(1)
```
Why do you use `sys.exit(1)` here? Wouldn't an exception have the same effect but be more specific and flexible? (Same at the bottom, of course.)

SimeonEhrig replied:
No, you call a normal application without required arguments, it say you that the arguments are missing. But it will not print the call stack and other technical information.

chillenzer replied:
Yes, that's okay.

chillenzer commented on share/ci/install/requirements_txt_modifier.py:
```diff
@@ -53,21 +53,27 @@ def cs(text: str, color: str):
 
-    with open(sys.argv[1], "r", encoding="utf-8") as input:
-        with open(sys.argv[2], "w", encoding="utf-8") as output:
+    with open(sys.argv[-2], "r", encoding="utf-8") as input:
+        with open(sys.argv[-1], "w", encoding="utf-8") as output:
```
This looks horribly dangerous. I'm aware that we're guarding against less than 3 arguments above but if the next person touching this file isn't aware of that, we might write the content of this very Python script into the output file. Maybe we could already extract the filenames in the else branch of the `if len(...` condition above, so there is at least proximity between those.

SimeonEhrig replied:
If I want make it save, I need to use `argparse`. I check the option.

chillenzer replied:
As I said, an intermediate step towards safety could be to extract those in the `if` condition.

chillenzer commented on share/ci/install/requirements_txt_modifier.py:
```diff
@@ -53,21 +53,27 @@ def cs(text: str, color: str):
     if len(sys.argv) < 3:
-        print("Set path to requirements.txt as first argument and output path as" " second argument.")
-        exit(1)
+        print("Set the path of the input requirements.txt as the second last argument "
+              "and the path of the output requirements.txt as the last argument.")
+        print("Set the name of the environment variables to ignore "
+              "before the path of the input requirements.txt")
```
Wouldn't we want to print them coloured as below?

chillenzer replied:
I think you wanted your other comment to appear here. I'll resolve this one.

chillenzer reviewed (changes requested):
Looks good but there are one or two things to discuss.

SimeonEhrig commented:
I didn't fixed the comments. I pushed again to trigger the CI to see, if the script is called somewhere else. I will fix the issues later.

SimeonEhrig commented:
@chillenzer Can you check it again. I think now it looks much better.

chillenzer commented on share/ci/install/pypicongpu.sh:
```diff
@@ -43,8 +43,17 @@ micromamba activate pypicongpu
 MODIFIED_REQUIREMENT_TXT_PYPICONGPU=$CI_PROJECT_DIR/lib/python/picongpu/pypicongpu/modified_requirements.txt
-python3 $CI_PROJECT_DIR/share/ci/install/requirements_txt_modifier.py $CI_PROJECT_DIR/lib/python/picongpu/picmi/requirements.txt $MODIFIED_REQUIREMENT_TXT_PICMI
-python3 $CI_PROJECT_DIR/share/ci/install/requirements_txt_modifier.py $CI_PROJECT_DIR/lib/python/picongpu/pypicongpu/requirements.txt $MODIFIED_REQUIREMENT_TXT_PYPICONGPU
+# The environment variables for changing the version of a dependency package are set globally.
+# Therefore, all calls of the requirements_txt_modfifier.py script are affected.
+# To avoid unexpected changes to dependencies, the list of ignored environment variables.
```
Very good! Thanks!

chillenzer commented on share/ci/install/requirements_txt_modifier.py:
```diff
@@ -52,35 +53,63 @@ def cs(text: str, color: str):
+        type=str,
+        nargs="*",
+        default=[],
+        help="Ignore this environment variables, which are set to modify the requirements.txt. The environment variables starts with `PYPIC_DEP_VERSION_`.",
```
```suggestion
        help="Ignore these environment variables, which are set to modify the requirements.txt. The environment variables starts with `PYPIC_DEP_VERSION_`.",
```

chillenzer reviewed (dismissed):
There are some minor things but I'd say nothing strictly blocking.

SimeonEhrig committed:
fix warning in PyPIConGPU version tests

- If the script `requirements_txt_modifier.py` should modify a dependency in a requirements.txt, which does not exist, it throws a warning because this is unexpected behavior. Because warnings are hard to detect in the CI, change it to an error.
- Implement a function to ignore specific environment variables which are set to change the version in a requirements.txt.

SimeonEhrig commented:
ping @chillenzer

chillenzer reviewed (approved):
Thought I already gave my approval but here it comes again.
